### ISSUE TITLE
Support dynamic imports

### ISSUE THREAD

**Issue by CodeWitchBella** [PepsRyuu/nollup#28]
Using dynamic imports produces following error:

```
{ SyntaxError: 'import' and 'export' may only appear at the top level (29:2)
    at _class.pp$4.raise (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:2825:15)
    at _class.pp$1.parseStatement (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:859:18)
    at _class.parseStatement (<project>/node_modules/acorn-dynamic-import/lib/index.js:63:118)
    at _class.pp$1.parseBlock (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:1161:23)
    at _class.pp$1.parseStatement (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:845:36)
    at _class.parseStatement (<project>/node_modules/acorn-dynamic-import/lib/index.js:63:118)
    at _class.pp$1.parseIfStatement (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:990:28)
    at _class.pp$1.parseStatement (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:834:33)
    at _class.parseStatement (<project>/node_modules/acorn-dynamic-import/lib/index.js:63:118)
    at _class.pp$1.parseTopLevel (<project>/node_modules/nollup/node_modules/acorn/dist/acorn.js:746:23)
  pos: 924,
  loc: Position { line: 29, column: 2 },
  raisedAt: 930 }
ParseError: src/index.tsx
Error: SyntaxError: 'import' and 'export' may only appear at the top level (29:2)
      import('react-axe')
     ^
    at bundle (<project>/node_modules/nollup/lib/index.js:393:19)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Possibly related to acornjs/acorn#809

**Comment by PepsRyuu** [comment]
Hey @CodeWitchBella! Thanks for reporting the issue. Can you provide a sample code snippet? 🙂

**Comment by PepsRyuu** [comment]
Can't seem to reproduce this issue. Does it happen in the ```example-dynamic-import``` app?

**Comment by CodeWitchBella** [comment]
Reproduction steps:

```
git clone https://github.com/CodeWitchBella/rollup-native-modules-boilerplate
cd rollup-native-modules-boilerplate 
npm ci
npx nollup -c
```

**Comment by CodeWitchBella** [comment]
Repo above produces slightly different error message but it still fails on dynamic import.

**Comment by PepsRyuu** [comment]
Thanks! It appears to be an Acorn and NPM resolution problem:

The ```node_modules``` folder has the following structure:
```
acorn/
acorn-dynamic-import/
nollup/
     acorn/
```
As far as I can tell, what's happening is that nollup requires its local acorn, which then requires the top-level acorn-dynamic-import, which then requires acorn. However, because acorn-dynamic-import is on the top-level, it requires the top-level acorn and not the local one. This is a huge problem, because plugins for Acorn are written like this:

```
          if (this.type === _acorn.tokTypes._import && parenAfter.call(this)) {
```
 Plugins mutate acorn and inject token types into it. For some reason, ```this.type``` and ```_acorn.tokTypes._import``` are objects, so it's doing a shallow comparison. Since nollup and acorn-dynamic-import are requiring two different versions of acorn, the comparison fails.

Thankfully there is a fix. I have to upgrade to the latest major version of Acorn which now includes dynamic imports in the core instead of relying on a plugin. Also need to implement ```emitFile``` which plugins you're using depends on. 🙂

I'll see what I can do and will release a new version as soon as possible!

**Comment by PepsRyuu** [comment]
I've released two new packages: ```nollup@0.7.0``` and ```rollup-plugin-commonjs-alternate@0.0.7```.

Nollup has been updated to include more plugin API methods (namely ```emitFile```) and to update Acorn so that it uses the native ```import()``` parsing. The CommonJS plugin has also been updated because ```hljs``` and ```lodash-es``` were using structures that were not being checked for. 

To get the skeleton working, the following steps need to be taken:

* Update Nollup and CommonJS plugin.
* Move ```replace()``` in ```rollup.config.js``` before ```commonjs()```. This allows the ```process.env.NODE_ENV``` check in React to work correctly. 
* Change ```import * as hljs``` to use default import - ```import hljs```. 
* Ideally change the ```process.env.NODE_ENV``` from ```production``` to ```development``` when using Nollup.
* The ```index.html``` file won't work because it uses a preprocessor that Nollup dev-server isn't aware of it. You'll need to create a separate index, with ```<script type="module" src="main-[hash].js"></script>```. When running nollup, make sure to pass ```--content-base public``` or the name of the directory that the index file resides in.

Personally I recommend having a look at ```example-react-hot-loader``` in this repository for the best way to use Nollup with React, so you can take full advantage of Hot Module Replacement.

Thanks again for reporting the issue, was helpful in identifying some very important bugs. Let me know if you run into any further problems! 🙂

**Comment by CodeWitchBella** [comment]
Wow. That was fast. Thank you 🙂
